sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala

```diff
@@ -88,6 +88,16 @@ abstract class BinaryArithmetic extends BinaryOperator {
 
   override def dataType: DataType = left.dataType
 
+  // Whether the dataType is promoted or not, we should only promote DecimalType once
+  var decimalPromoted: Boolean = false
```

Comment:
The HiveTypeCoercion will cast the children to increase the precision and scale, this should be done only once. Currently, we rely on turning them into Unlimited decimal.

I haven't figure out a better to do it, anyway.